Tests/DeepLearningTests/TrivialModelTests.swift

```diff
@@ -40,8 +40,8 @@ final class TrivialModelTests: XCTestCase {
                 return l2.applied(to: h1, in: context)
             }
         }
-        let optimizer = SGD<Classifier, Float>(learningRate: 0.02)
         var classifier = Classifier(hiddenSize: 4)
+        let optimizer = SGD(learningRate: 0.02, for: classifier)
```

Comment:
There's actually a much more serious issue: Have you actually tested it to make sure this really works?

I don't think this would train at all. Floating point literal `0.02` defaults to `Double`, so the `Scalar` associated type in `SGD` would be double. Then it'll try to get key paths to `Tensor<Double>`, which do not exist at all in a model where all parameters are `Tensor<Float>`. So my hypothesis is that this will compile but definitely fail to train. This also exposes how confusing it would be to not make optimizers take a generic type or not make the model provide a currency scalar type.

Potential solutions:
1. Make the `Layer` protocol actually require a `Scalar` associated type as the currency type, but that would be unideal because it would trigger a lot of protocol conformance errors when the user didn't specify a `Scalar` type alias.
2. Still require a scalar type explicitly in the optimizer. The way to do that is to require a "scalarType" metatype argument as before: When it's specified as an argument, everything generic parameter gets inferred; when it's not specified as an argument, the compiler would expect it to be provided as a generic type parameter. The crucial thing here is to make `Scalar` *not* be inferred from the `learningRate:` argument -- I thought about this earlier, and this is why I added an explicit `scalarType:` argument in the first place.

Reply:
I agree with solution 2: with the current design, scalar type should be made explicit for optimizers.

Reply:
Yup, you're right that what happens is that it doesn't train. (I didn't notice because it compiled, and there was a previously failing test.)

I've filed [TF-302](https://bugs.swift.org/browse/TF-302) to follow up with the larger usability issue, and fixed up the API and tests so they're all passing now.